pageserver/src/deletion_queue.rs

```diff
@@ -0,0 +1,1299 @@
+            &self.tx,
+            FrontendQueueMessage::Delete(DeletionOp {
+                tenant_id,
+                timeline_id,
+                layers,
+                generation: current_generation,
+                objects: Vec::new(),
+            }),
+        )
+        .await
+    }
+
+    async fn do_flush<T>(
+        &self,
+        queue: &tokio::sync::mpsc::Sender<T>,
+        msg: T,
+        rx: tokio::sync::oneshot::Receiver<()>,
+    ) -> Result<(), DeletionQueueError> {
+        self.do_push(queue, msg).await?;
+        if rx.await.is_err() {
+            // This shouldn't happen if tenants are shut down before deletion queue.  If we
+            // encounter a bug like this, then a flusher will incorrectly believe it has flushed
+            // when it hasn't, possibly leading to leaking objects.
+            error!("Deletion queue dropped flush op while client was still waiting");
+            Err(DeletionQueueError::ShuttingDown)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Wait until all previous deletions are persistent (either executed, or written to a DeletionList)
+    pub async fn flush(&self) -> Result<(), DeletionQueueError> {
+        let (tx, rx) = tokio::sync::oneshot::channel::<()>();
+        self.do_flush(&self.tx, FrontendQueueMessage::Flush(FlushOp { tx }), rx)
+            .await
+    }
+
+    // Wait until all previous deletions are executed
+    pub(crate) async fn flush_execute(&self) -> Result<(), DeletionQueueError> {
+        debug!("flush_execute: flushing to deletion lists...");
+        // Flush any buffered work to deletion lists
+        self.flush().await?;
+
+        // Flush the backend into the executor of deletion lists
+        let (tx, rx) = tokio::sync::oneshot::channel::<()>();
+        debug!("flush_execute: flushing backend...");
+        self.do_flush(
+            &self.tx,
+            FrontendQueueMessage::FlushExecute(FlushOp { tx }),
+            rx,
+        )
+        .await?;
+        debug!("flush_execute: finished flushing backend...");
+
+        // Flush any immediate-mode deletions (the above backend flush will only flush
+        // the executor if deletions had flowed through the backend)
+        debug!("flush_execute: flushing execution...");
+        let (tx, rx) = tokio::sync::oneshot::channel::<()>();
+        self.do_flush(
+            &self.executor_tx,
+            ExecutorMessage::Flush(FlushOp { tx }),
+            rx,
+        )
+        .await?;
+        debug!("flush_execute: finished flushing execution...");
+        Ok(())
+    }
+
+    /// This interface bypasses the persistent deletion queue, and any validation
+    /// that this pageserver is still elegible to execute the deletions.  It is for
+    /// use in timeline deletions, where the control plane is telling us we may
+    /// delete everything in the timeline.
+    ///
+    /// DO NOT USE THIS FROM GC OR COMPACTION CODE.  Use the regular `push_layers`.
```

Comment:
With #4938 callsites for layer deletion will become callsite-independent. This constraint will be impossible to uphold.

Reply:
Oki, it's fine to still use schedule_layer_file_deletion in remote timeline client.

Reply:
It's totally legal to use schedule_layer_file_deletion, but a bit inefficient for timeline deletions, because they don't need any of the RemoteTimelineClient ordering logic, or the validation of deletes in the deletion queue backend.

The clean answer to this will probably be https://github.com/neondatabase/neon/issues/5080 -- so that timeline deletions just don't touch any of the runtime Tenant/Layer/RemoteTimelineClient machinery that they don't need, and just drive the remote storage interface to delete the objects, which actually might not be very complicated to do, so we can probably do it in the near future.

Reply:
Oki, sounds good, resolving.